apollo-core/src/main/java/com/ctrip/framework/foundation/internals/NetworkInterfaceManager.java

```diff
@@ -97,27 +97,13 @@ private void load() {
     }
 
     try {
-      Enumeration<NetworkInterface> interfaces = NetworkInterface.getNetworkInterfaces();
-      List<NetworkInterface> nis = interfaces == null ? Collections.<NetworkInterface>emptyList() : Collections.list(interfaces);
-      List<InetAddress> addresses = new ArrayList<InetAddress>();
-      InetAddress local = null;
-
-      try {
-        for (NetworkInterface ni : nis) {
-          if (ni.isUp() && !ni.isLoopback()) {
-            addresses.addAll(Collections.list(ni.getInetAddresses()));
-          }
-        }
-        local = findValidateIp(addresses);
-      } catch (Exception e) {
-        // ignore
-      }
-      if (local != null) {
-        m_local = local;
+      final InetAddress localHost = InetAddress.getLocalHost();
```

Comment:
This change might change the original behavior, have you tested this change in a multi-network interface environment?

Reply:
yes, I tested this on my mac

[~] ifconfig                                                                                                                                             15:53:06  ☁  master ☂ ⚡ ✚ ✭
lo0: flags=8049<UP,LOOPBACK,RUNNING,MULTICAST> mtu 16384
	options=1203<RXCSUM,TXCSUM,TXSTATUS,SW_TIMESTAMP>
	inet 127.0.0.1 netmask 0xff000000
	inet6 ::1 prefixlen 128
	inet6 fe80::1%lo0 prefixlen 64 scopeid 0x1
	nd6 options=201<PERFORMNUD,DAD>
gif0: flags=8010<POINTOPOINT,MULTICAST> mtu 1280
stf0: flags=0<> mtu 1280
en0: flags=8863<UP,BROADCAST,SMART,RUNNING,SIMPLEX,MULTICAST> mtu 1500
	ether f4:5c:89:ae:6f:cd
	inet6 fe80::c26:ade8:4fba:ee27%en0 prefixlen 64 secured scopeid 0x4
	inet 192.168.3.6 netmask 0xffffff00 broadcast 192.168.3.255
	nd6 options=201<PERFORMNUD,DAD>
	media: autoselect
	status: active
en1: flags=963<UP,BROADCAST,SMART,RUNNING,PROMISC,SIMPLEX> mtu 1500
	options=60<TSO4,TSO6>
	ether 4a:00:05:5e:19:e0
	media: autoselect <full-duplex>
	status: inactive
en2: flags=963<UP,BROADCAST,SMART,RUNNING,PROMISC,SIMPLEX> mtu 1500
	options=60<TSO4,TSO6>
	ether 4a:00:05:5e:19:e1
	media: autoselect <full-duplex>
	status: inactive
bridge0: flags=8863<UP,BROADCAST,SMART,RUNNING,SIMPLEX,MULTICAST> mtu 1500
	options=63<RXCSUM,TXCSUM,TSO4,TSO6>
	ether 4a:00:05:5e:19:e0
	Configuration:
		id 0:0:0:0:0:0 priority 0 hellotime 0 fwddelay 0
		maxage 0 holdcnt 0 proto stp maxaddr 100 timeout 1200
		root id 0:0:0:0:0:0 priority 0 ifcost 0 port 0
		ipfilter disabled flags 0x2
	member: en1 flags=3<LEARNING,DISCOVER>
	        ifmaxaddr 0 port 5 priority 0 path cost 0
	member: en2 flags=3<LEARNING,DISCOVER>
	        ifmaxaddr 0 port 6 priority 0 path cost 0
	nd6 options=201<PERFORMNUD,DAD>
	media: <unknown type>
	status: inactive
p2p0: flags=8843<UP,BROADCAST,RUNNING,SIMPLEX,MULTICAST> mtu 2304
	ether 06:5c:89:ae:6f:cd
	media: autoselect
	status: inactive
awdl0: flags=8943<UP,BROADCAST,RUNNING,PROMISC,SIMPLEX,MULTICAST> mtu 1484
	ether ea:92:a5:3b:47:13
	inet6 fe80::e892:a5ff:fe3b:4713%awdl0 prefixlen 64 scopeid 0x9
	nd6 options=201<PERFORMNUD,DAD>
	media: autoselect
	status: active
utun0: flags=8051<UP,POINTOPOINT,RUNNING,MULTICAST> mtu 2000
	inet6 fe80::7d1e:25ff:e7e2:4c7a%utun0 prefixlen 64 scopeid 0xa
	nd6 options=201<PERFORMNUD,DAD>
vnic0: flags=8843<UP,BROADCAST,RUNNING,SIMPLEX,MULTICAST> mtu 1500
	options=3<RXCSUM,TXCSUM>
	ether 00:1c:42:00:00:08
	inet 10.211.55.2 netmask 0xffffff00 broadcast 10.211.55.255
	media: autoselect
	status: active
vnic1: flags=8843<UP,BROADCAST,RUNNING,SIMPLEX,MULTICAST> mtu 1500
	options=3<RXCSUM,TXCSUM>
	ether 00:1c:42:00:00:09
	inet 10.37.129.2 netmask 0xffffff00 broadcast 10.37.129.255
	media: autoselect
	status: active
vboxnet0: flags=8843<UP,BROADCAST,RUNNING,SIMPLEX,MULTICAST> mtu 1500
	ether 0a:00:27:00:00:00
	inet 192.168.99.1 netmask 0xffffff00 broadcast 192.168.99.255

test method
public static void main(String[] args) {
    String value = null;

    value = System.getProperty("host.ip");
    System.out.println("1 \"host.ip\" "+value);

    if (value == null) {
      value = System.getenv("host.ip");
      System.out.println("2 \"host.ip\" "+value);
    }

//    NetworkInterface.getNetworkInterfaces();
//    NetworkInterface.

    try {




      Enumeration<NetworkInterface> interfaces = NetworkInterface.getNetworkInterfaces();
      List<NetworkInterface> nis = interfaces == null ? Collections.<NetworkInterface>emptyList() : Collections.list(interfaces);
      final List<InetAddress> addresses = new ArrayList<InetAddress>();
      InetAddress local = null;

      try {
        for (NetworkInterface ni : nis) {
          if (ni.isUp() && !ni.isLoopback()&& !ni.isVirtual()&&!ni.isPointToPoint()) {
            final Enumeration<InetAddress> inetAddresses = ni.getInetAddresses();

            addresses.addAll(Collections.list(inetAddresses));
          }
        }
        for (InetAddress address:
                addresses) {
//          if (address.)
          System.out.println(address.getHostAddress());//ip
          System.out.println(address.getHostName());//主机名
          System.out.println(address.getCanonicalHostName());//主机名
          System.out.println("本地主机ip  "+address.getLocalHost().getHostAddress());//本地主机ip【果然最有用。。】
        }
      } catch (Exception e) {
        // ignore
      }

    } catch (SocketException e) {
      // ignore it
    }


    System.out.println();
    DefaultNetworkProvider defaultNetworkProvider = new DefaultNetworkProvider();
    System.out.println(defaultNetworkProvider);
  }

the test output
1 "host.ip" null
2 "host.ip" null
fe80:0:0:0:e892:a5ff:fe3b:4713%awdl0
fe80:0:0:0:e892:a5ff:fe3b:4713%awdl0
fe80:0:0:0:e892:a5ff:fe3b:4713%awdl0
本地主机ip  192.168.3.6
192.168.99.1
192.168.99.1
192.168.99.1
本地主机ip  192.168.3.6
10.37.129.2
10.37.129.2
10.37.129.2
本地主机ip  192.168.3.6
10.211.55.2
10.211.55.2
10.211.55.2
本地主机ip  192.168.3.6
fe80:0:0:0:c26:ade8:4fba:ee27%en0
johnsons-nmbp.local
johnsons-nmbp.local
本地主机ip  192.168.3.6
192.168.3.6
192.168.3.6
192.168.3.6
本地主机ip  192.168.3.6

Reply:
I need to find more environments to test this one...

Reply:
I just tested in one of our environment, and it turns out the result of InetAddress.getLocalHost() is not the same as Foundation.net().getHostAddress().

After some investigation, the main reason is in our environments, we have a host mapping in /etc/hosts, like the following(someHostName is the real host name of the machine)

```
127.0.0.1    someHostName
```

which in this case, InetAddress.getLocalHost().getHostAddress() will always return 127.0.0.1 while Foundation.net().getHostAddress() returns the correct address.

Please understand that I'm not saying `InetAddress.getLocalHost().getHostAddress()` is wrong, it's just not correct in such environment setup..

So for this open source project, I might have to leave this part of code as it is, though it's kind of slow...

Reply:
i understand your concern, for a open source product like Apollo, your point is right, but on my env, i also setup '127.0.0.1	localhost peer1 peer2',  but i get the correct result, i also test on my vm, the result as follow:

➜  classes git:(johnson) ✗ java com.ctrip.framework.foundation.internals.provider.DefaultNetworkProvider
1 "host.ip" null
2 "host.ip" null
fe80:0:0:0:482d:fff:febf:7ae9%docker0
fe80:0:0:0:482d:fff:febf:7ae9%docker0
fe80:0:0:0:482d:fff:febf:7ae9%docker0
本地主机ip  10.211.55.9
172.17.42.1
172.17.42.1
172.17.42.1
本地主机ip  10.211.55.9
fe80:0:0:0:21c:42ff:fe39:aa50%eth0
fe80:0:0:0:21c:42ff:fe39:aa50%eth0
fe80:0:0:0:21c:42ff:fe39:aa50%eth0
本地主机ip  10.211.55.9
fdb2:2c26:f4e4:0:21c:42ff:fe39:aa50%eth0
^C%                                                                                                                                                                                  ➜  classes git:(johnson) ✗ cat /etc/hosts
127.0.0.1   quill.net localhost localhost.localdomain localhost4 localhost4.localdomain4
::1         quill.net localhost localhost.localdomain localhost6 localhost6.localdomain6
10.211.55.7 centoslocal1 CentOS-6-Mini-1
10.211.55.9 centoslocal0 CentOS-6-Mini-0
10.211.55.10 centoslocal2 CentOS-6-Mini-2

so, can you test for other environment, i mean not only ctrip?

Reply:
dubbo-common/src/main/java/com/alibaba/dubbo/common/utils/NetUtils.java

FYI. the dubbo also will cause some error, apollo also will cause some error, how to balance， your duty！^_^  but, if let me choose, i choose InetAddress.getLocalHost().getHostAddress(), because it's fast..

Reply:
can you run the following commands on your mac and attach your result?

nslookup 192.168.3.6

nslookup 192.168.99.1

I'm thinking we might be able to combine your idea with the original algorithm, such as to replace the following 

```java
        // has host name
        // TODO fix performance issue when calling getHostName
        if (!Objects.equals(address.getHostName(), address.getHostAddress())) {
          weight += 1;
        }
```

to 

```java
        if (Objects.equals(InetAddress.getLocalHost(), address)) {
          weight += 1;
        }
```

Reply:
I have conducted more tests to cover more scenarios.

## Network interfaces( interfaces like p2p0, utun0 are omitted for simplicity ): 

```
lo0: flags=8049<UP,LOOPBACK,RUNNING,MULTICAST> mtu 16384
        options=1203<RXCSUM,TXCSUM,TXSTATUS,SW_TIMESTAMP>
        inet 127.0.0.1 netmask 0xff000000 
        inet6 ::1 prefixlen 128 
        inet6 fe80::1%lo0 prefixlen 64 scopeid 0x1 
        nd6 options=201<PERFORMNUD,DAD>
en4: flags=8863<UP,BROADCAST,SMART,RUNNING,SIMPLEX,MULTICAST> mtu 1500
        options=4<VLAN_MTU>
        inet 10.32.21.4 netmask 0xffffff00 broadcast 10.32.21.255
        nd6 options=201<PERFORMNUD,DAD>
        media: autoselect (100baseTX <full-duplex,flow-control>)
        status: active
```

## Test program:

```java
System.out.println("Foundation.net().getHostAddress(): " + Foundation.net().getHostAddress());
System.out.println("InetAddress.getLocalHost().getHostAddress(): " + InetAddress.getLocalHost().getHostAddress());
```

1. Case 1 - Run the test program directly
```
Foundation.net().getHostAddress(): 10.32.21.4
InetAddress.getLocalHost().getHostAddress(): 10.32.21.22
```

The result of `InetAddress.getLocalHost().getHostAddress()` is clearly wrong, there might be something wrong in our gateway(since nslookup $myHostName returns 10.32.21.22), but the result shows the output of `InetAddress.getLocalHost().getHostAddress()` could be some ip which has no connection to my computer!

2. Case 2 - Edit /etc/hosts to add `10.32.21.4 $myHostName` and run the test program
>$myHostName is just a placeholder, which should be replaced by the result of executing `hostname` command

```
Foundation.net().getHostAddress(): 10.32.21.4
InetAddress.getLocalHost().getHostAddress(): 10.32.21.4
```

This time, both outputs are correct. 

So the output of `InetAddress.getLocalHost().getHostAddress()` is the ip of either the result of nslookup $myHostName, or the one configured in /etc/hosts. The latter one takes precedence.

3. Case 3 - Edit /etc/hosts to add `add 127.0.0.1 $myHostName` and run the test program
>$myHostName is just a placeholder, which should be replaced by the result of executing `hostname` command

```
Foundation.net().getHostAddress(): 10.32.21.4
InetAddress.getLocalHost().getHostAddress(): 127.0.0.1
```

This is not surprising based on the test result of case 2.

4. Case 4 - Edit /etc/hosts to add `add 1.2.3.4 $myHostName` and run the test program
>$myHostName is just a placeholder, which should be replaced by the result of executing `hostname` command

```
Foundation.net().getHostAddress(): 10.32.21.4
InetAddress.getLocalHost().getHostAddress(): 1.2.3.4
```

Again, this is not surprising based on the test result of case 2 and case 3.

## Conclusion

The result of `InetAddress.getLocalHost().getHostAddress()` is not reliable, since it depends on the result of `nslookup` and the config in `/etc/hosts`, which in worst case, could return an ip that has no connection to the computer itself!

Foundation.net().getHostAddress() could be slow if there are many network interfaces in the computer(only the first time), but it won't return an ip which doesn't belong to the computer itself.

It might return an ip address which is not the best (since all the network interfaces' weights are the same), but we could always adjust the weight by adding `ip $hostName` in the `/etc/hosts` so that we could have the correct ip address returned, which is the way we are using in ctrip.